diff --git a/test/test-thread-priority.c b/test/test-thread-priority.c
index 0aaf2977..c93caee9 100644
--- a/test/test-thread-priority.c
+++ b/test/test-thread-priority.c
@@ -61,11 +61,11 @@ TEST_IMPL(thread_priority) {
   ASSERT_OK(uv_thread_create(&task_id, simple_task, NULL));
   ASSERT_OK(uv_thread_getpriority(task_id, &priority));
 
-#ifdef _WIN32
-  ASSERT_EQ(priority, THREAD_PRIORITY_NORMAL);
+#if defined(_WIN32) || defined(__MVS__)
+  ASSERT_EQ(priority, UV_THREAD_PRIORITY_NORMAL);
 #else
   ASSERT_OK(pthread_getschedparam(task_id, &policy, &param));
-#ifdef __PASE__
+#if defined(__PASE__) || defined(__MVS__)
   min = 1;
   max = 127;
 #else
@@ -102,4 +102,4 @@ TEST_IMPL(thread_priority) {
   uv_sem_destroy(&sem);
 
   return 0;
-}
\ No newline at end of file
+}
