diff --git a/test/test-thread-priority.c b/test/test-thread-priority.c
index acc33855..33d8b99f 100644
--- a/test/test-thread-priority.c
+++ b/test/test-thread-priority.c
@@ -45,6 +45,10 @@ static void simple_task(void *args) {
 }
 
 TEST_IMPL(thread_priority) {
+#ifdef __MVS__
+  RETURN_SKIP("Thread priority cannot be changed on z/OS.");
+  return 0;
+#else
   int priority;
 #ifndef _WIN32
   int min;
@@ -106,4 +110,5 @@ TEST_IMPL(thread_priority) {
 #endif
 
   return 0;
+#endif // __MVS__
 }
