diff --git a/test/test-fs-copyfile.c b/test/test-fs-copyfile.c
index f7a0c236..c5c9c884 100644
--- a/test/test-fs-copyfile.c
+++ b/test/test-fs-copyfile.c
@@ -49,6 +49,10 @@ static void handle_result(uv_fs_t* req) {
   uint64_t uid;
   uint64_t gid;
   int r;
+#ifdef __MVS__
+  struct stat src_st;
+  struct stat dst_st;
+#endif
 
   ASSERT_EQ(req->fs_type, UV_FS_COPYFILE);
   ASSERT_OK(req->result);
@@ -61,6 +65,15 @@ static void handle_result(uv_fs_t* req) {
   uid = stat_req.statbuf.st_uid;
   gid = stat_req.statbuf.st_gid;
   uv_fs_req_cleanup(&stat_req);
+#ifdef __MVS__
+/* Verify that file tags are the same. */
+  r = stat(req->path, &src_st);
+  ASSERT_EQ(r, 0);
+  r = stat(dst, &dst_st);
+  ASSERT_EQ(r, 0);
+  ASSERT_EQ(src_st.st_tag.ft_txtflag, dst_st.st_tag.ft_txtflag);
+  ASSERT_EQ(src_st.st_tag.ft_ccsid, dst_st.st_tag.ft_ccsid);
+#endif
   r = uv_fs_stat(NULL, &stat_req, dst, NULL);
   ASSERT_OK(r);
   ASSERT_EQ(stat_req.statbuf.st_size, size);
