diff --git a/test/test-fs.c b/test/test-fs.c
index 4761b15b..cd357fec 100644
--- a/test/test-fs.c
+++ b/test/test-fs.c
@@ -3039,9 +3039,9 @@ TEST_IMPL(fs_lutime) {
   atime = mtime = 400497753.25; /* 1982-09-10 11:22:33.25 */
 
   r = uv_fs_lutime(NULL, &req, symlink_path, atime, mtime, NULL);
-#if (defined(_AIX) && !defined(_AIX71)) || defined(__MVS__)
+#if (defined(_AIX) && !defined(_AIX71)) 
   ASSERT_EQ(r, UV_ENOSYS);
-  RETURN_SKIP("lutime is not implemented for z/OS and AIX versions below 7.1");
+  RETURN_SKIP("lutime is not implemented for AIX versions below 7.1");
 #endif
   ASSERT_OK(r);
   ASSERT_OK(req.result);
