diff --git a/src/unix/tcp.c b/src/unix/tcp.c
index a6b53e59..7fed0846 100644
--- a/src/unix/tcp.c
+++ b/src/unix/tcp.c
@@ -36,7 +36,9 @@
 #define getifaddrs Qp2getifaddrs
 #define freeifaddrs Qp2freeifaddrs
 #else
-#include <ifaddrs.h>
+#ifndef __MVS__
+ #include <ifaddrs.h>
+#endif
 #endif
 
 static int maybe_bind_socket(int fd) {
@@ -225,10 +227,13 @@ static int uv__is_ipv6_link_local(const struct sockaddr* addr) {
 
 static int uv__ipv6_link_local_scope_id(void) {
   struct sockaddr_in6* a6;
-  struct ifaddrs* ifa;
-  struct ifaddrs* p;
   int rv;
 
+#ifdef __MVS__
+  return 0;
+#else
+  struct ifaddrs* ifa;
+  struct ifaddrs* p;
   if (getifaddrs(&ifa))
     return 0;
 
@@ -244,6 +249,7 @@ static int uv__ipv6_link_local_scope_id(void) {
 
   freeifaddrs(ifa);
   return rv;
+#endif
 }
 
 
