diff --git a/src/unix/tcp.c b/src/unix/tcp.c
index 799fca77..b24003bc 100644
--- a/src/unix/tcp.c
+++ b/src/unix/tcp.c
@@ -30,8 +30,8 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 
-/* ifaddrs is not implemented on AIX and IBM i PASE */
-#if !defined(_AIX)
+/* ifaddrs is not implemented on AIX and IBM i PASE and z/OS */
+#if !defined(_AIX) && !defined(__MVS__)
 #include <ifaddrs.h>
 #endif
 
@@ -222,7 +222,7 @@ static int uv__is_ipv6_link_local(const struct sockaddr* addr) {
 static int uv__ipv6_link_local_scope_id(void) {
   struct sockaddr_in6* a6;
   int rv;
-#if defined(_AIX)
+#if defined(_AIX) || defined(__MVS__)
   /* AIX & IBM i do not have ifaddrs
    * so fallback to use uv_interface_addresses */
   uv_interface_address_t* interfaces;
@@ -262,7 +262,7 @@ static int uv__ipv6_link_local_scope_id(void) {
   }
 
   freeifaddrs(ifa);
-#endif /* defined(_AIX) */
+#endif /* defined(_AIX) && !defined(__MVS__) */
 
   return rv;
 }
