diff --git a/src/unix/tcp.c b/src/unix/tcp.c
index e4e2e028..c591428b 100644
--- a/src/unix/tcp.c
+++ b/src/unix/tcp.c
@@ -29,7 +29,9 @@
 
 #include <sys/types.h>
 #include <sys/socket.h>
+#ifndef __MVS__
 #include <ifaddrs.h>
+#endif
 
 static int maybe_bind_socket(int fd) {
   union uv__sockaddr s;
@@ -217,10 +219,13 @@ static int uv__is_ipv6_link_local(const struct sockaddr* addr) {
 
 static int uv__ipv6_link_local_scope_id(void) {
   struct sockaddr_in6* a6;
-  struct ifaddrs* ifa;
-  struct ifaddrs* p;
   int rv;
 
+#ifdef __MVS__
+  return 0;
+#else
+  struct ifaddrs* ifa;
+  struct ifaddrs* p;
   if (getifaddrs(&ifa))
     return 0;
 
@@ -236,6 +241,7 @@ static int uv__ipv6_link_local_scope_id(void) {
 
   freeifaddrs(ifa);
   return rv;
+#endif
 }


