diff --git a/test/test-fs-copyfile.c b/test/test-fs-copyfile.c
index 3aacf125..12fb7ed5 100644
--- a/test/test-fs-copyfile.c
+++ b/test/test-fs-copyfile.c
@@ -47,6 +47,10 @@ static void handle_result(uv_fs_t* req) {
   uint64_t size;
   uint64_t mode;
   int r;
+#ifdef __MVS__
+  struct stat src_st;
+  struct stat dst_st;
+#endif
 
   ASSERT_EQ(req->fs_type, UV_FS_COPYFILE);
   ASSERT_OK(req->result);
@@ -57,6 +61,15 @@ static void handle_result(uv_fs_t* req) {
   size = stat_req.statbuf.st_size;
   mode = stat_req.statbuf.st_mode;
   uv_fs_req_cleanup(&stat_req);
+#ifdef __MVS__
+  /* Verify that file tags are the same. */
+  r = stat(req->path, &src_st);
+  ASSERT_EQ(r, 0);
+  r = stat(dst, &dst_st);
+  ASSERT_EQ(r, 0);
+  ASSERT_EQ(src_st.st_tag.ft_txtflag, dst_st.st_tag.ft_txtflag);
+  ASSERT_EQ(src_st.st_tag.ft_ccsid, dst_st.st_tag.ft_ccsid);
+#endif
   r = uv_fs_stat(NULL, &stat_req, dst, NULL);
   ASSERT_OK(r);
   ASSERT_EQ(stat_req.statbuf.st_size, size);
