diff --git a/CMakeLists.txt b/CMakeLists.txt
index 73d5aff8..9b47d172 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -365,23 +365,12 @@ if(CMAKE_SYSTEM_NAME STREQUAL "OS390")
        src/unix/os390-syscalls.c
        src/unix/os390-proctitle.c)
   list(APPEND uv_cflags
-       -q64
-       -qascii
-       -qexportall
-       -qgonumber
-       -qlongname
-       -qlibansi
-       -qfloat=IEEE
-       -qtune=10
-       -qarch=10
-       -qasm
-       -qasmlib=sys1.maclib:sys1.modgen)
-  find_library(ZOSLIB
-    NAMES zoslib
-    PATHS ${ZOSLIB_DIR}
-    PATH_SUFFIXES lib
-  )
-  list(APPEND uv_libraries ${ZOSLIB})
+    -fgnu-keywords
+    -fno-short-enums
+    -fzos-le-char-mode=ascii
+    -m64
+    -mzos-target=zosv2r4
+    -march=arch10)
 endif()
 
 if(CMAKE_SYSTEM_NAME STREQUAL "OS400")
@@ -469,6 +458,12 @@ if(LIBUV_BUILD_SHARED)
   if(CMAKE_SYSTEM_NAME STREQUAL "OS390")
     target_include_directories(uv PUBLIC $<BUILD_INTERFACE:${ZOSLIB_DIR}/include>)
     set_target_properties(uv PROPERTIES LINKER_LANGUAGE CXX)
+    find_library(ZOSLIB
+      NAMES "libzoslib.x"
+      PATHS ${ZOSLIB_DIR}
+      PATH_SUFFIXES "lib"
+    )
+    target_link_libraries(uv ${ZOSLIB})
   endif()
   target_link_libraries(uv ${uv_libraries})
   set_target_properties(uv PROPERTIES OUTPUT_NAME "uv")
@@ -486,6 +481,12 @@ target_include_directories(uv_a
 if(CMAKE_SYSTEM_NAME STREQUAL "OS390")
   target_include_directories(uv_a PUBLIC $<BUILD_INTERFACE:${ZOSLIB_DIR}/include>)
   set_target_properties(uv_a PROPERTIES LINKER_LANGUAGE CXX)
+  find_library(ZOSLIB_A
+    NAMES "libzoslib.a"
+    PATHS ${ZOSLIB_DIR}
+    PATH_SUFFIXES "lib"
+  )
+  target_link_libraries(uv_a ${ZOSLIB_A})
 endif()
 target_link_libraries(uv_a ${uv_libraries})
 set_target_properties(uv_a PROPERTIES OUTPUT_NAME "uv")
