diff --git a/src/unix/internal.h b/src/unix/internal.h
index 80fe41f8..40b4205d 100644
--- a/src/unix/internal.h
+++ b/src/unix/internal.h
@@ -99,6 +99,14 @@
 # define UV__PATH_MAX 8192
 #endif
 
+#ifndef S_ISDIR
+# if defined(_S_IFMT) && defined(_S_IFDIR)
+#  define S_ISDIR(a)   (((a) & _S_IFMT) == _S_IFDIR)
+# else
+#  define S_ISDIR(a)   (((a) & S_IFMT) == S_IFDIR)
+# endif
+#endif
+
 union uv__sockaddr {
   struct sockaddr_in6 in6;
   struct sockaddr_in in;
@@ -305,7 +313,7 @@ void uv__udp_io(uv_loop_t* loop, uv__io_t* w, unsigned int events);
 #define uv__fs_event(loop, w, events) UNREACHABLE()
 #endif
 
-#ifndef __linux__
+#if !defined(__linux__) && !defined(__MVS__)
 #define uv__inotify_read(loop, w, events) UNREACHABLE()
 #endif
 
