diff --git a/src/unix/internal.h b/src/unix/internal.h
index bcb3be57..e606cf97 100644
--- a/src/unix/internal.h
+++ b/src/unix/internal.h
@@ -92,6 +92,14 @@
 # define UV__PATH_MAX 8192
 #endif
 
+#ifndef S_ISDIR
+# if defined(_S_IFMT) && defined(_S_IFDIR)
+#  define S_ISDIR(a)   (((a) & _S_IFMT) == _S_IFDIR)
+# else
+#  define S_ISDIR(a)   (((a) & S_IFMT) == S_IFDIR)
+# endif
+#endif
+
 union uv__sockaddr {
   struct sockaddr_in6 in6;
   struct sockaddr_in in;
