diff --git a/src/unix/linux.c b/src/unix/linux.c
index 763f5dd5..9560aaa5 100644
--- a/src/unix/linux.c
+++ b/src/unix/linux.c
@@ -489,6 +489,8 @@ static int uv__use_io_uring(uint32_t flags) {
   use = atomic_load_explicit(&use_io_uring, memory_order_relaxed);
 
   if (use == 0) {
+  /* Disable io_uring by default due to CVE-2024-22017. */
+    use = -1;
     val = getenv("UV_USE_IO_URING");
     use = val != NULL && atoi(val) > 0 ? 1 : -1;
     atomic_store_explicit(&use_io_uring, use, memory_order_relaxed);
@@ -499,6 +501,14 @@ static int uv__use_io_uring(uint32_t flags) {
 }
 
 
+UV_EXTERN int uv__node_patch_is_using_io_uring(void) {
+  // This function exists only in the modified copy of libuv in the Node.js
+  // repository. Node.js checks if this function exists and, if it does, uses it
+  // to determine whether libuv is using io_uring or not.
+  return uv__use_io_uring();
+}
+
+
 static void uv__iou_init(int epollfd,
                          struct uv__iou* iou,
                          uint32_t entries,
