diff --git a/src/unix/os390.c b/src/unix/os390.c
index 1b277292..b5ccc1c0 100644
--- a/src/unix/os390.c
+++ b/src/unix/os390.c
@@ -19,19 +19,23 @@
  * IN THE SOFTWARE.
  */
 
+/*
+ * inotify code in this file is based on uv/src/unix/linux.c
+ */
+
 #include "uv.h"
 #include "internal.h"
 #include <sys/ioctl.h>
 #include <net/if.h>
 #include <utmpx.h>
 #include <unistd.h>
+#include <sys/inotify.h>
 #include <sys/ps.h>
 #include <builtins.h>
 #include <termios.h>
 #include <sys/msg.h>
 #include <sys/resource.h>
 #include "zos-base.h"
-#include "zos-sys-info.h"
 #if defined(__clang__)
 #include "csrsic.h"
 #else
@@ -80,13 +84,6 @@
 */
 #define RAXFMCT           0x2C
 
-/* Thread Entry constants */
-#define PGTH_CURRENT  1
-#define PGTH_LEN      26
-#define PGTHAPATH     0x20
-#pragma linkage(BPX4GTH, OS)
-#pragma linkage(BPX1GTH, OS)
-
 /* TOD Clock resolution in nanoseconds */
 #define TOD_RES 4.096
 
@@ -114,18 +111,32 @@ void uv_loadavg(double avg[3]) {
 int uv__platform_loop_init(uv_loop_t* loop) {
   uv__os390_epoll* ep;
 
-  ep = epoll_create1(0);
-  loop->ep = ep;
+  // Init these early in case __epoll_create1 failed.
+  loop->inotify_watchers = NULL;
+  loop->inotify_fd = -1;
+  loop->ep = NULL;
+  loop->backend_fd = -1;
+
+  ep = __epoll_create1(0);
   if (ep == NULL)
     return UV__ERR(errno);
 
+  loop->ep = ep;
+  loop->backend_fd = ep->msg_queue;
+
   return 0;
 }
 
 
 void uv__platform_loop_delete(uv_loop_t* loop) {
+  if (loop->inotify_fd != -1) {
+    uv__io_stop(loop, &loop->inotify_read_watcher, POLLIN);
+    uv__close(loop->inotify_fd);
+    loop->inotify_fd = -1;
+  }
   if (loop->ep != NULL) {
     epoll_queue_close(loop->ep);
+    uv__free(loop->ep);
     loop->ep = NULL;
   }
 }
@@ -155,7 +166,6 @@ static int getexe(char* buf, size_t len) {
 int uv_exepath(char* buffer, size_t* size) {
   int res;
   char args[PATH_MAX];
-  int pid;
 
   if (buffer == NULL || size == NULL || *size == 0)
     return UV_EINVAL;
@@ -189,11 +199,15 @@ uint64_t uv_get_total_memory(void) {
 uint64_t uv_get_constrained_memory(void) {
   struct rlimit rl;
 
-  /* RLIMIT_MEMLIMIT return value is in megabytes rather than bytes. */
-  if (getrlimit(RLIMIT_MEMLIMIT, &rl) == 0)
+  /* RLIMIT_MEMLIMIT return value is in megabytes rather than bytes.
+   * https://www.ibm.com/docs/en/zos/3.1.0?topic=functions-getrlimit-get-current-maximum-resource-consumption
+   * If > RLIM_INFINITY, user limit likely set to unlimited.
+   * Otherwise there is no memory limit set if getrlimit failed.
+   */
+  if (getrlimit(RLIMIT_MEMLIMIT, &rl) == 0 && rl.rlim_cur < RLIM_INFINITY)
     return rl.rlim_cur * 1024 * 1024;
 
-  return 0; /* There is no memory limit set. */
+  return 0;
 }
 
 
@@ -389,6 +403,7 @@ static int uv__interface_addresses_v6(uv_interface_address_t** addresses,
     address->netmask.netmask6.sin6_len = p->__nif6e_prefixlen;
     memcpy(&(address->netmask.netmask6.sin6_addr), netmask, 16);
     address->netmask.netmask6.sin6_family = AF_INET6;
+    address->address.address6.sin6_scope_id = if_nametoindex(address->name);
 
     address->is_internal = p->__nif6e_flags & _NIF6E_FLAGS_LOOPBACK ? 1 : 0;
     address++;
@@ -585,15 +600,15 @@ void uv_free_interface_addresses(uv_interface_address_t* addresses,
 
 
 void uv__platform_invalidate_fd(uv_loop_t* loop, int fd) {
-  struct epoll_event* events;
-  struct epoll_event dummy;
+  struct __epoll_event* events;
+  struct __epoll_event dummy;
   uintptr_t i;
   uintptr_t nfds;
 
   assert(loop->watchers != NULL);
   assert(fd >= 0);
 
-  events = (struct epoll_event*) loop->watchers[loop->nwatchers];
+  events = (struct __epoll_event*) loop->watchers[loop->nwatchers];
   nfds = (uintptr_t) loop->watchers[loop->nwatchers + 1];
   if (events != NULL)
     /* Invalidate events with same file descriptor */
@@ -602,8 +617,11 @@ void uv__platform_invalidate_fd(uv_loop_t* loop, int fd) {
         events[i].fd = -1;
 
   /* Remove the file descriptor from the epoll. */
+  dummy.events = 0;
+  dummy.fd = fd;
+  dummy.is_msg = 0;
   if (loop->ep != NULL)
-    epoll_ctl(loop->ep, EPOLL_CTL_DEL, fd, &dummy);
+    __epoll_ctl(loop->ep, EPOLL_CTL_DEL, fd, &dummy);
 }
 
 
@@ -633,114 +651,341 @@ int uv_fs_event_init(uv_loop_t* loop, uv_fs_event_t* handle) {
   return 0;
 }
 
+STATIC_ASSERT(EPOLL_CTL_ADD < 4);
+STATIC_ASSERT(EPOLL_CTL_DEL < 4);
+STATIC_ASSERT(EPOLL_CTL_MOD < 4);
+
+struct watcher_list {
+  RB_ENTRY(watcher_list) entry;
+  struct uv__queue watchers;
+  int iterating;
+  char* path;
+  int wd;
+};
+
+struct watcher_root {
+  struct watcher_list* rbh_root;
+};
+
+static int uv__inotify_fork(uv_loop_t* loop, struct watcher_list* root);
+static void uv__inotify_read(uv_loop_t* loop,
+                             uv__io_t* w,
+                             unsigned int revents);
+static int compare_watchers(const struct watcher_list* a,
+                            const struct watcher_list* b);
+static void maybe_free_watcher_list(struct watcher_list* w,
+                                    uv_loop_t* loop);
+
+RB_GENERATE_STATIC(watcher_root, watcher_list, entry, compare_watchers)
+
+
+static struct watcher_root* uv__inotify_watchers(uv_loop_t* loop) {
+  /* This cast works because watcher_root is a struct with a pointer as its
+   * sole member. Such type punning is unsafe in the presence of strict
+   * pointer aliasing (and is just plain nasty) but that is why libuv
+   * is compiled with -fno-strict-aliasing.
+   */
+  return (struct watcher_root*) &loop->inotify_watchers;
+}
+
+
+int uv__io_fork(uv_loop_t* loop) {
+  int err;
+  struct watcher_list* root;
+
+  root = uv__inotify_watchers(loop)->rbh_root;
+
+  uv__close(loop->backend_fd);
+  loop->backend_fd = -1;
+
+  uv__platform_loop_delete(loop);
+
+  /*
+    Nullify the msg queue but don't close it because
+    it is still being used by the parent.
+  */
+  loop->ep = NULL;
+
+  err = uv__platform_loop_init(loop);
+  if (err)
+    return err;
+
+  return uv__inotify_fork(loop, root);
+}
 
-static int os390_regfileint(uv_fs_event_t* handle, char* path) {
-  uv__os390_epoll* ep;
-  _RFIS reg_struct;
-  int rc;
 
-  ep = handle->loop->ep;
-  assert(ep->msg_queue != -1);
+static int compare_watchers(const struct watcher_list* a,
+                            const struct watcher_list* b) {
+  if (a->wd < b->wd) return -1;
+  if (a->wd > b->wd) return 1;
+  return 0;
+}
 
-  reg_struct.__rfis_cmd  = _RFIS_REG;
-  reg_struct.__rfis_qid  = ep->msg_queue;
-  reg_struct.__rfis_type = 1;
-  memcpy(reg_struct.__rfis_utok, &handle, sizeof(handle));
 
-  rc = __w_pioctl(path, _IOCC_REGFILEINT, sizeof(reg_struct), &reg_struct);
-  if (rc != 0)
+static int init_inotify(uv_loop_t* loop) {
+  int err;
+  int fd;
+
+  if (loop->inotify_fd != -1)
+    return 0;
+
+  fd = inotify_init1(IN_NONBLOCK | IN_CLOEXEC);
+  if (fd < 0)
     return UV__ERR(errno);
 
-  memcpy(handle->rfis_rftok, reg_struct.__rfis_rftok,
-         sizeof(handle->rfis_rftok));
+  err = uv__io_init_start(loop, &loop->inotify_read_watcher, uv__inotify_read,
+                          fd, POLLIN);
+  if (err) {
+    uv__close(fd);
+    return err;
+  }
 
+  loop->inotify_fd = fd;
   return 0;
 }
 
 
-int uv_fs_event_start(uv_fs_event_t* handle, uv_fs_event_cb cb,
-                      const char* filename, unsigned int flags) {
-  char* path;
-  int rc;
+static int uv__inotify_fork(uv_loop_t* loop, struct watcher_list* root) {
+  /* Open the inotify_fd, and re-arm all the inotify watchers. */
+  int err;
+  struct watcher_list* tmp_watcher_list_iter;
+  struct watcher_list* watcher_list;
+  struct watcher_list tmp_watcher_list;
+  struct uv__queue queue;
+  struct uv__queue* q;
+  uv_fs_event_t* handle;
+  char* tmp_path;
 
-  if (uv__is_active(handle))
-    return UV_EINVAL;
+  if (root == NULL)
+    return 0;
 
-  path = uv__strdup(filename);
-  if (path == NULL)
-    return UV_ENOMEM;
+  /* We must restore the old watcher list to be able to close items
+   * out of it.
+   */
+  loop->inotify_watchers = root;
 
-  rc = os390_regfileint(handle, path);
-  if (rc != 0) {
-    uv__free(path);
-    return rc;
+  uv__queue_init(&tmp_watcher_list.watchers);
+  /* Note that the queue we use is shared with the start and stop()
+   * functions, making uv__queue_foreach unsafe to use. So we use the
+   * uv__queue_move trick to safely iterate. Also don't free the watcher
+   * list until we're done iterating. c.f. uv__inotify_read.
+   */
+  RB_FOREACH_SAFE(watcher_list, watcher_root,
+                  uv__inotify_watchers(loop), tmp_watcher_list_iter) {
+    watcher_list->iterating = 1;
+    uv__queue_move(&watcher_list->watchers, &queue);
+    while (!uv__queue_empty(&queue)) {
+      q = uv__queue_head(&queue);
+      handle = uv__queue_data(q, uv_fs_event_t, watchers);
+      /* It's critical to keep a copy of path here, because it
+       * will be set to NULL by stop() and then deallocated by
+       * maybe_free_watcher_list
+       */
+      tmp_path = uv__strdup(handle->path);
+      assert(tmp_path != NULL);
+      uv__queue_remove(q);
+      uv__queue_insert_tail(&watcher_list->watchers, q);
+      uv_fs_event_stop(handle);
+
+      uv__queue_insert_tail(&tmp_watcher_list.watchers, &handle->watchers);
+      handle->path = tmp_path;
+    }
+    watcher_list->iterating = 0;
+    maybe_free_watcher_list(watcher_list, loop);
   }
 
-  uv__handle_start(handle);
-  handle->path = path;
-  handle->cb = cb;
+  uv__queue_move(&tmp_watcher_list.watchers, &queue);
+  while (!uv__queue_empty(&queue)) {
+      q = uv__queue_head(&queue);
+      uv__queue_remove(q);
+      handle = uv__queue_data(q, uv_fs_event_t, watchers);
+      tmp_path = handle->path;
+      handle->path = NULL;
+      err = uv_fs_event_start(handle, handle->cb, tmp_path, 0);
+      uv__free(tmp_path);
+      if (err)
+        return err;
+  }
 
   return 0;
 }
 
 
-int uv__fs_event_stop(uv_fs_event_t* handle) {
-  uv__os390_epoll* ep;
-  _RFIS reg_struct;
-  int rc;
+static struct watcher_list* find_watcher(uv_loop_t* loop, int wd) {
+  struct watcher_list w;
+  w.wd = wd;
+  return RB_FIND(watcher_root, uv__inotify_watchers(loop), &w);
+}
 
-  if (!uv__is_active(handle))
-    return 0;
 
-  ep = handle->loop->ep;
-  assert(ep->msg_queue != -1);
+static void maybe_free_watcher_list(struct watcher_list* w, uv_loop_t* loop) {
+  /* if the watcher_list->watchers is being iterated over, we can't free it. */
+  if ((!w->iterating) && uv__queue_empty(&w->watchers)) {
+    /* No watchers left for this path. Clean up. */
+    RB_REMOVE(watcher_root, uv__inotify_watchers(loop), w);
+    inotify_rm_watch(loop->inotify_fd, w->wd);
+    uv__free(w);
+  }
+}
 
-  reg_struct.__rfis_cmd  = _RFIS_UNREG;
-  reg_struct.__rfis_qid  = ep->msg_queue;
-  reg_struct.__rfis_type = 1;
-  memcpy(reg_struct.__rfis_rftok, handle->rfis_rftok,
-         sizeof(handle->rfis_rftok));
 
-  /*
-   * This call will take "/" as the path argument in case we
-   * don't care to supply the correct path. The system will simply
-   * ignore it.
-   */
-  rc = __w_pioctl("/", _IOCC_REGFILEINT, sizeof(reg_struct), &reg_struct);
-  if (rc != 0 && errno != EALREADY && errno != ENOENT)
-    abort();
+static void uv__inotify_read(uv_loop_t* loop,
+                             uv__io_t* dummy,
+                             unsigned int events) {
+  const struct inotify_event* e;
+  struct watcher_list* w;
+  uv_fs_event_t* h;
+  struct uv__queue queue;
+  struct uv__queue* q;
+  const char* path;
+  ssize_t size;
+  const char *p;
+  /* needs to be large enough for sizeof(inotify_event) + strlen(path) */
+  char buf[4096];
+
+  for (;;) {
+    do
+      size = read(loop->inotify_fd, buf, sizeof(buf));
+    while (size == -1 && errno == EINTR);
+
+    if (size == -1) {
+      assert(errno == EAGAIN || errno == EWOULDBLOCK);
+      break;
+    }
+
+    assert(size > 0); /* pre-2.6.21 thing, size=0 == read buffer too small */
+
+    /* Now we have one or more inotify_event structs. */
+    for (p = buf; p < buf + size; p += sizeof(*e) + e->len) {
+      e = (const struct inotify_event*) p;
+
+      events = 0;
+      if (e->mask & (IN_ATTRIB|IN_MODIFY))
+        events |= UV_CHANGE;
+      if (e->mask & ~(IN_ATTRIB|IN_MODIFY))
+        events |= UV_RENAME;
+
+      w = find_watcher(loop, e->wd);
+      if (w == NULL)
+        continue; /* Stale event, no watchers left. */
 
-  if (handle->path != NULL) {
-    uv__free(handle->path);
-    handle->path = NULL;
+      /* inotify does not return the filename when monitoring a single file
+       * for modifications. Repurpose the filename for API compatibility.
+       * I'm not convinced this is a good thing, maybe it should go.
+       */
+      path = e->len ? (const char*) (e + 1) : uv__basename_r(w->path);
+
+      /* We're about to iterate over the queue and call user's callbacks.
+       * What can go wrong?
+       * A callback could call uv_fs_event_stop()
+       * and the queue can change under our feet.
+       * So, we use uv__queue_move() trick to safely iterate over the queue.
+       * And we don't free the watcher_list until we're done iterating.
+       *
+       * First,
+       * tell uv_fs_event_stop() (that could be called from a user's callback)
+       * not to free watcher_list.
+       */
+      w->iterating = 1;
+      uv__queue_move(&w->watchers, &queue);
+      while (!uv__queue_empty(&queue)) {
+        q = uv__queue_head(&queue);
+        h = uv__queue_data(q, uv_fs_event_t, watchers);
+
+        uv__queue_remove(q);
+        uv__queue_insert_tail(&w->watchers, q);
+
+        h->cb(h, path, events, 0);
+      }
+      /* done iterating, time to (maybe) free empty watcher_list */
+      w->iterating = 0;
+      maybe_free_watcher_list(w, loop);
+    }
   }
+}
 
-  if (rc != 0 && errno == EALREADY)
-    return -1;
 
-  uv__handle_stop(handle);
+int uv_fs_event_start(uv_fs_event_t* handle,
+                      uv_fs_event_cb cb,
+                      const char* path,
+                      unsigned int flags) {
+  struct watcher_list* w;
+  uv_loop_t* loop;
+  size_t len;
+  int events;
+  int err;
+  int wd;
+
+  if (uv__is_active(handle))
+    return UV_EINVAL;
+
+  loop = handle->loop;
+
+  err = init_inotify(loop);
+  if (err)
+    return err;
+
+  events = IN_ATTRIB
+         | IN_CREATE
+         | IN_MODIFY
+         | IN_DELETE
+         | IN_DELETE_SELF
+         | IN_MOVE_SELF
+         | IN_MOVED_FROM
+         | IN_MOVED_TO;
+
+  wd = inotify_add_watch(loop->inotify_fd, path, events);
+  if (wd == -1)
+    return UV__ERR(errno);
+
+  w = find_watcher(loop, wd);
+  if (w)
+    goto no_insert;
+
+  len = strlen(path) + 1;
+  w = uv__malloc(sizeof(*w) + len);
+  if (w == NULL)
+    return UV_ENOMEM;
+
+  w->wd = wd;
+  w->path = memcpy(w + 1, path, len);
+  uv__queue_init(&w->watchers);
+  w->iterating = 0;
+  RB_INSERT(watcher_root, uv__inotify_watchers(loop), w);
+
+no_insert:
+  uv__handle_start(handle);
+  uv__queue_insert_tail(&w->watchers, &handle->watchers);
+  handle->path = w->path;
+  handle->cb = cb;
+  handle->wd = wd;
 
   return 0;
 }
 
 
 int uv_fs_event_stop(uv_fs_event_t* handle) {
-  uv__fs_event_stop(handle);
+  struct watcher_list* w;
+
+  if (!uv__is_active(handle))
+    return 0;
+
+  w = find_watcher(handle->loop, handle->wd);
+  assert(w != NULL);
+
+  handle->wd = -1;
+  handle->path = NULL;
+  uv__handle_stop(handle);
+  uv__queue_remove(&handle->watchers);
+
+  maybe_free_watcher_list(w, handle->loop);
+
   return 0;
 }
 
 
 void uv__fs_event_close(uv_fs_event_t* handle) {
-  /*
-   * If we were unable to unregister file interest here, then it is most likely
-   * that there is a pending queued change notification. When this happens, we
-   * don't want to complete the close as it will free the underlying memory for
-   * the handle, causing a use-after-free problem when the event is processed.
-   * We defer the final cleanup until after the event is consumed in
-   * os390_message_queue_handler().
-   */
-  if (uv__fs_event_stop(handle) == 0)
-    uv__make_close_pending((uv_handle_t*) handle);
+  uv_fs_event_stop(handle);
 }
 
 
@@ -755,29 +1000,53 @@ static int os390_message_queue_handler(uv__os390_epoll* ep) {
 
   msglen = msgrcv(ep->msg_queue, &msg, sizeof(msg), 0, IPC_NOWAIT);
 
-  if (msglen == -1 && errno == ENOMSG)
-    return 0;
-
-  if (msglen == -1)
-    abort();
+  if (msglen == -1) {
+    if (errno == ENOMSG)
+      return 0;
+    else if (errno == EINVAL)
+      return -1;
+    else
+      abort();
+  }
 
   events = 0;
   if (msg.__rfim_event == _RFIM_ATTR || msg.__rfim_event == _RFIM_WRITE)
     events = UV_CHANGE;
   else if (msg.__rfim_event == _RFIM_RENAME || msg.__rfim_event == _RFIM_UNLINK)
     events = UV_RENAME;
-  else if (msg.__rfim_event == 156)
-    /* TODO(gabylb): zos - this event should not happen, need to investigate.
+  else if (msg.__rfim_event == 156) {
+    /* TODO(gabylb): still needed? */
+    /* TODO(gabylb): investigate why this event is occuring as it should
+     * not happen based on https://github.ibm.com/open-z/node-build/issues/199.
+     *
+     * This (undocumented?) event seems to occur when the watched file
+     * is [re]moved, or an eeitor (like vim) renames then creates the file
+     * on save (for vim, that's when backupcopy=no|auto), which causes
+     * __w_pioctl to fail as the file wasn't yet created; the sleep is
+     * to help resolve the rename-create issue.
      *
-     * This event seems to occur when the watched file is [re]moved, or an
-     * editor (like vim) renames then creates the file on save (for vim, that's
-     * when backupcopy=no|auto).
+     * This is a temporary workaround because on extremely busy systems,
+     * timeouts cannot guarantee deterministic behaviour. Since the inode
+     * number changes, they are technically NOT the same file. Thus this kind
+     * of monitoring is not really valid. If there is a real issue, the
+     * application itself should be adjusted to deal with time gap. The current
+     * monitoring method has a gap, when inotify becomes availlable, we should
+     * switch to it.
+     *
+     * See: https://github.ibm.com/open-z/node-build/issues/656
+     *
+     * TODO: zos - resolve this issue on z/OS for V2R4 if possible, and
+     * remove entirely when V2R5 is the min supported level and only inotify
+     * is used. This should not be upstreamed.
      */
+    struct timespec timeout = { 0, 1000000 }; /* 1ms. */
     events = UV_RENAME;
-  else
+    nanosleep(&timeout, NULL);
+  } else
     /* Some event that we are not interested in. */
     return 0;
 
+  /* TODO(gabylb): still needed? */
   /* `__rfim_utok` is treated as text when it should be treated as binary while
    * running in ASCII mode, resulting in an unwanted autoconversion.
    */
@@ -788,7 +1057,9 @@ static int os390_message_queue_handler(uv__os390_epoll* ep) {
   assert((handle->flags & UV_HANDLE_CLOSED) == 0);
   if (uv__is_closing(handle)) {
     uv__handle_stop(handle);
-    uv__make_close_pending((uv_handle_t*) handle);
+    if (access(handle->path, F_OK) == 0) {
+       uv__make_close_pending((uv_handle_t*) handle);
+    }
     return 0;
   } else if (handle->path == NULL) {
     /* _RFIS_UNREG returned EALREADY. */
@@ -796,13 +1067,13 @@ static int os390_message_queue_handler(uv__os390_epoll* ep) {
     return 0;
   }
 
+  /* TODO(gabylb): still needed? */
   /* The file is implicitly unregistered when the change notification is
    * sent, only one notification is sent per registration. So we need to
    * re-register interest in a file after each change notification we
    * receive.
    */
   assert(handle->path != NULL);
-  os390_regfileint(handle, handle->path);
   handle->cb(handle, uv__basename_r(handle->path), events, 0);
   return 1;
 }
@@ -811,9 +1082,9 @@ static int os390_message_queue_handler(uv__os390_epoll* ep) {
 void uv__io_poll(uv_loop_t* loop, int timeout) {
   static const int max_safe_timeout = 1789569;
   uv__loop_internal_fields_t* lfields;
-  struct epoll_event events[1024];
-  struct epoll_event* pe;
-  struct epoll_event e;
+  struct __epoll_event events[1024];
+  struct __epoll_event* pe;
+  struct __epoll_event e;
   uv__os390_epoll* ep;
   int have_signals;
   int real_timeout;
@@ -836,8 +1107,6 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
   lfields = uv__get_internal_fields(loop);
 
   while (!uv__queue_empty(&loop->watcher_queue)) {
-    uv_stream_t* stream;
-
     q = uv__queue_head(&loop->watcher_queue);
     uv__queue_remove(q);
     uv__queue_init(q);
@@ -845,13 +1114,11 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
 
     assert(w->pevents != 0);
     assert(w->fd >= 0);
-
-    stream= container_of(w, uv_stream_t, io_watcher);
-
     assert(w->fd < (int) loop->nwatchers);
 
     e.events = w->pevents;
     e.fd = w->fd;
+    e.is_msg = 0;
 
     if (w->events == 0)
       op = EPOLL_CTL_ADD;
@@ -861,14 +1128,14 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
     /* XXX Future optimization: do EPOLL_CTL_MOD lazily if we stop watching
      * events, skip the syscall and squelch the events after epoll_wait().
      */
-    if (epoll_ctl(loop->ep, op, w->fd, &e)) {
+    if (__epoll_ctl(loop->ep, op, w->fd, &e)) {
       if (errno != EEXIST)
         abort();
 
       assert(op == EPOLL_CTL_ADD);
 
       /* We've reactivated a file descriptor that's been watched before. */
-      if (epoll_ctl(loop->ep, EPOLL_CTL_MOD, w->fd, &e))
+      if (__epoll_ctl(loop->ep, EPOLL_CTL_MOD, w->fd, &e))
         abort();
     }
 
@@ -907,8 +1174,8 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
      */
     lfields->current_timeout = timeout;
 
-    nfds = epoll_wait(loop->ep, events,
-                      ARRAY_SIZE(events), timeout);
+    nfds = __epoll_wait(loop->ep, events,
+                        ARRAY_SIZE(events), timeout);
 
     /* Update loop->time unconditionally. It's tempting to skip the update when
      * timeout == 0 (i.e. non-blocking poll) but there is no guarantee that the
@@ -970,7 +1237,13 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
 
       ep = loop->ep;
       if (pe->is_msg) {
-        os390_message_queue_handler(ep);
+        if (os390_message_queue_handler(ep) == -1) {
+          /* The user has deleted the System V message queue. Highly likely
+           * that the process is being shut down. So stop listening to it.
+           */
+          __epoll_ctl(loop->ep, EPOLL_CTL_DEL, ep->msg_queue, pe);
+          loop->backend_fd = -1;
+        }
         nevents++;
         continue;
       }
@@ -986,7 +1259,7 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
          * Ignore all errors because we may be racing with another thread
          * when the file descriptor is closed.
          */
-        epoll_ctl(loop->ep, EPOLL_CTL_DEL, fd, pe);
+        __epoll_ctl(loop->ep, EPOLL_CTL_DEL, fd, pe);
         continue;
       }
 
@@ -1058,13 +1331,3 @@ update_timeout:
   }
 }
 
-
-int uv__io_fork(uv_loop_t* loop) {
-  /*
-    Nullify the msg queue but don't close it because
-    it is still being used by the parent.
-  */
-  loop->ep = NULL;
-
-  return uv__platform_loop_init(loop);
-}
