diff --git a/src/unix/pipe.c b/src/unix/pipe.c
index 68e225e2..5a0410eb 100644
--- a/src/unix/pipe.c
+++ b/src/unix/pipe.c
@@ -473,8 +473,9 @@ int uv_pipe_chmod(uv_pipe_t* handle, int mode) {
     desired_mode |= S_IWUSR | S_IWGRP | S_IWOTH;
 
   /* fchmod on macOS and (Free|Net|Open)BSD does not support UNIX sockets. */
+  /* fchmod on z/OS fails with errno == ENOSYS */
   if (fchmod(fd, desired_mode))
-    if (errno != EINVAL && errno != EOPNOTSUPP)
+    if (errno != EINVAL && errno != EOPNOTSUPP && errno != ENOSYS)
       return UV__ERR(errno);
 
   /* Fall back to chmod. */
