diff --git a/src/unix/signal.c b/src/unix/signal.c
index ccaa72db..90a6d4c4 100644
--- a/src/unix/signal.c
+++ b/src/unix/signal.c
@@ -183,9 +183,11 @@ static uv_signal_t* uv__signal_first_handle(int signum) {
 static void uv__signal_handler(int signum) {
   uv__signal_msg_t msg;
   uv_signal_t* handle;
-  int saved_errno;
-
-  saved_errno = errno;
+  int saved_errno = errno;
+#ifdef __MVS__
+  // This is required in case the signal was thrown from a non-ASCII env:
+  __ae_thread_swapmode(__AE_ASCII_MODE);
+#endif
   memset(&msg, 0, sizeof msg);
 
   if (uv__signal_lock()) {
