diff --git a/src/unix/tcp.c b/src/unix/tcp.c
index 799fca77..3ab8ce6a 100644
--- a/src/unix/tcp.c
+++ b/src/unix/tcp.c
@@ -30,8 +30,8 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 
-/* ifaddrs is not implemented on AIX and IBM i PASE */
-#if !defined(_AIX)
+/* ifaddrs is not implemented on AIX and z/OS */
+#if !defined(_AIX) && !defined(__MVS__)
 #include <ifaddrs.h>
 #endif
 
@@ -220,14 +220,13 @@ static int uv__is_ipv6_link_local(const struct sockaddr* addr) {
 
 
 static int uv__ipv6_link_local_scope_id(void) {
-  struct sockaddr_in6* a6;
   int rv;
-#if defined(_AIX)
+#if defined(_AIX) || defined(__MVS__)
   /* AIX & IBM i do not have ifaddrs
    * so fallback to use uv_interface_addresses */
   uv_interface_address_t* interfaces;
   uv_interface_address_t* ifa;
-  int count, i;
+  int count;
 
   if (uv_interface_addresses(&interfaces, &count))
     return 0;
@@ -244,6 +243,7 @@ static int uv__ipv6_link_local_scope_id(void) {
   uv_free_interface_addresses(interfaces, count);
 
 #else
+  struct sockaddr_in6* a6;
   struct ifaddrs* ifa;
   struct ifaddrs* p;
 
